### DISCUSSION TITLE
Setup questions for camera with Pi 3B

### DISCUSSION THREAD

**Discussion by tgkell** [iRobotEducation/create3_docs#90] [Questions] [answered]
I have a Pi 3B with Ubuntu Server installed, and it can see the ROS2 topics over the wireless network but not when wired to the Create with the setup explained below.

My naive plan is to have the ROS2 communication between the PI and the Create on the wired 192.168.186.0/24 network.  I would then communicate with the Pi on its wireless interface.  Question: is that a recommended setup?  I will be leading activities with about 30 students who are to each have their own Create, and I am exploring what we can do each in our own homes vs together in a classroom.  I don't have control over the wireless in the classroom, and I'm prepared to have only one Create going in the classroom at a time.

My current status: the Pi can see the ROS2 topics when the communication is wireless and the Pi is not wired to the Create.  When I set things up as follows, ROS2 stops showing all topics. I've included a command trace so you can see what is happening, and here's a summary of what's in that trace:

Hardware setup as in the Pi 3B Hookup Example, using the same USB-Ethernet adapter that is in the picture.
RMW implementation is Cyclone (both create and ubuntu)
CYCLONEDDS_URI environment variable points to a file containing the Cyclone config (see command trace) to use eth0 only
eth0 on Ubuntu is 192.168.186.3, and can ping the Create at 192.168.186.2
"ros2 topic list" shows just the two topics

Note: there is a second wired Ubuntu machine on the LAN with ROS2 installed but I'm assuming it's not interfering.
When I unplug the ethernet cable and remove the CYCLONEDDS_URI variable, and reboot, "ros2 topic list" works again.

Two Questions: 
1. Am I doing this wrong?
[command_trace.txt](https://github.com/iRobotEducation/create3_docs/files/8980605/command_trace.txt)

2. What is the recommended setup so that a camera can be added to the Create?  I'm thinking the Pi with camera will control the Create (wired) and the Pi will interact with a laptop wirelessly. But maybe the laptop should get a wireless camera stream from the Pi, and the laptop should control the create wirelessly?

**Comment by alsora** [comment, marked as answer]
Your proposed setup with the Pi controlling the create is definitely the recommended one to develop an autonomous robotics application.

We had reports of users having problems with this setup when using Cyclone dds.

You could:
 - make sure that you have the `DontRoute`option in your Cyclone xml files
 - try switching DDS implementation to Fast-DDS which seems more reliable in handling multiple interfaces (remember to change RMW on all 3 machines, you can do it on the robot via the webserver)


I recommend you to have a look at this documentation on RMW configuration https://iroboteducation.github.io/create3_docs/setup/xml-config/

**Reply by justinIRBT** [reply]
Also note that for this use case, you will want to setup an NTP configuration on your Pi to serve time to the Create3 if it doesn't have network access to keep the clocks synced between Create3 and Pi:
https://iroboteducation.github.io/create3_docs/setup/compute-ntp/

**Reply by tgkell** [reply]
Excuse the noob questions:

1. I am imagining in this setup (Pi with camera wired to Create) that only the Pi and Create will participate in ROS2 pub/sub.  The laptop is just connected to the Pi over SSH to tell the Pi what program to run, or stop, etc.  Is that not what should happen?  (alsora mentioned RMW on all three machines)
2. Followup to Question 1: how do I configure the Create to do ROS2 over the wired network with Pi only?  Take away its wifi password?

Regarding NTP when Create wifi is disabled, yes, point taken.  In my setup for my original post, wifi (but not BLE) is still there on the Create (isn't it?) and I was thinking that the Create might broadcast DDS on also wifi, but was hoping the Pi would see DDS on its eth0.  I will try switching to Fast-DDS.

**Reply by justinIRBT** [reply]
1) If you are uncertain if the wifi network configuration will work for ROS2, you should be able to use the system without it with the workflow you describe.  You can use CLI tools on the pi such as ros2 topic and ros2 action to debug and start the system.

2) If the create3’s wifi doesn’t get an IP address, but has a usb0 connection, it will automatically configure itself to just communicate over usb0.  If it gets an IP address on wifi sometime later on, it will restart the application configured to talk on both interfaces (this is why the process starts multiple times, because it is restarting when getting a wifi ip).  If you want your robot to forget about it’s network provisioning, you can use “factory reset” as described here:
https://iroboteducation.github.io/create3_docs/webserver/about/

The NTP setup is mostly important if the Create isn’t connected to wifi so it can keep it’s time synced with the Pi.

**Reply by tgkell** [reply]
Thanks!  I did a Create factory reset to ensure the Pi and Create are not using wifi, set the RMW to Fast-DDS on the Create (to match the Pi) and ros2 topic list works.
